src/main/scala/beam/analysis/LinkTraversalAnalysis.scala

```diff
@@ -0,0 +1,229 @@
+import scala.collection.mutable
+
+/**
+  * Analyze the link traversal events.
+  * @param scenario beam scenario.
+  * @param beamServices beam services.
+  * @param outputDirectoryHierarchy output directory hierarchy.
+  */
+class LinkTraversalAnalysis @Inject()(
+  private val scenario: Scenario,
+  private val beamServices: BeamServices,
+  outputDirectoryHierarchy: OutputDirectoryHierarchy
+) extends BasicEventHandler
+    with OutputDataDescriptor {
+
+  private val outputFileBaseName = "linkTraversalAnalysis"
+  private var analysisData: Array[LinkTraversalData] = Array()
+
+  /**
+    * Handles the PathTraversalEvent notification and generates the analysis data
+    * @param event Event
+    */
+  override def handleEvent(event: Event): Unit = {
+    try {
+      event match {
+        case pe if event.isInstanceOf[PathTraversalEvent] =>
+          val eventAttributes: mutable.Map[String, String] = pe.getAttributes.asScala
+          val linkIds = eventAttributes.getOrElse(PathTraversalEvent.ATTRIBUTE_LINK_IDS, "").split(",")
+          val linkTravelTimes: Array[String] = event.asInstanceOf[PathTraversalEvent].getLinkTravelTimes.split(",")
+          val vehicleId = eventAttributes.getOrElse(PathTraversalEvent.ATTRIBUTE_VEHICLE_ID, "")
+          val vehicleType = eventAttributes.getOrElse(PathTraversalEvent.ATTRIBUTE_VEHICLE_TYPE, "")
+          val pathArrivalTime = try {
+            eventAttributes.getOrElse(PathTraversalEvent.ATTRIBUTE_ARRIVAL_TIME, "").toLong
+          } catch {
+            case _: Exception => 0
+          }
+          val linkArrivalTimes: Seq[Long] = for (i <- linkTravelTimes.indices) yield {
+            i match {
+              case 0 => pathArrivalTime
+              case _ =>
+                pathArrivalTime + (try {
+                  linkTravelTimes(i - 1).toLong
+                } catch {
+                  case _: Exception => 0
+                })
+            }
+          }
+          val networkLinks = scenario.getNetwork.getLinks.values().asScala
+          val nextLinkIds = linkIds.toList.takeRight(linkIds.size - 1)
+          if (linkIds.nonEmpty) {
+            val data = linkIds zip nextLinkIds zip linkTravelTimes zip linkArrivalTimes flatMap { tuple =>
+              val (((id, nextId), travelTime), arrivalTime) = tuple
+              val nextLink: Option[Link] = networkLinks.find(x => x.getId.toString.equals(nextId))
+              networkLinks.find(x => x.getId.toString.equals(id)) map { currentLink =>
+                val freeFlowSpeed = currentLink.getFreespeed
+                val linkCapacity = currentLink.getCapacity
+                val averageSpeed = try {
+                  currentLink.getLength / travelTime.toInt
+                } catch {
+                  case _: Exception => 0.0
+                }
+                val turnAtLinkEnd =
+                  getDirection(currentLink.getCoord, nextLink.map(_.getCoord).getOrElse(new Coord(0.0, 0.0)))
```

Comment:
Oops, I didn't catch this before. This is not the correct math to calculate the turning angle. What this does is effectively calculate the cardinal direction. 

What we want is to pass 2 "coords" which actually represent direction vectors of each link.

So the first argument to "getDirection" should be vectorFromLink(currentLink) and second argument vectorFromLink(nextLink.map(_.getCoord).getOrElse(currentLink))

***note how I used currentLink as the default, this is means that if the nextLink is empty we end up with a Straight instead of some unanticipated turn when you pick an arbitrary 0 coordinate.***

where vectorFromLink(link: Link)

new Coord(link.getToNode.getCoord.getX - link.getFromNode.getCoord.getX, link.getToNode.getCoord.getY - link.getFromNode.getCoord.getY)